adapters/sp1/groth16-verifier/src/types/vk.rs

```diff
@@ -0,0 +1,100 @@
+use bn::{AffineG2, G2};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    error::{Error, Groth16Error},
+    types::{
+        g1::{compressed_bytes_to_affine_g1, SAffineG1},
+        g2::{compressed_bytes_to_affine_g2, SAffineG2},
+    },
+};
+
+/// G1 elements of the verification key.
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub(crate) struct Groth16G1 {
+    pub(crate) alpha: SAffineG1,
+    pub(crate) k: Vec<SAffineG1>,
+}
+
+/// G2 elements of the verification key.
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub(crate) struct Groth16G2 {
+    pub(crate) beta: SAffineG2,
+    pub(crate) delta: SAffineG2,
+    pub(crate) gamma: SAffineG2,
+}
+
+/// Verification key for the Groth16 proof.
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub(crate) struct Groth16VerifyingKey {
+    pub(crate) g1: Groth16G1,
+    pub(crate) g2: Groth16G2,
+}
+
+impl Groth16VerifyingKey {
+    /// Load a Groth16 verifying key from the GNARK‐style compressed byte slice.
```

Comment:
Recommend expanding this comment to detail the required structure of the byte slice, since there's a lot of element extraction via specific indexing in this function that is not immediately apparent.

Reply:
This will also panic on a buffer that is too short, which should at the very least be documented (if not prevented entirely).